adapter/internal/oasparser/model/policy_container.go

```diff
@@ -0,0 +1,251 @@
+	// using index i instead of struct to validate policy against multiple allowed and apply only first if multiple exists
+	spec := p[policy.PolicyName].Specification
+	if err := spec.validatePolicy(policy, flow, stats, index); err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Operation policy validation failed, ignoring the policy \"%v\": %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2204,
+		})
+		return policy, err
+	}
+
+	defRaw := p[policy.PolicyName].Definition.RawData
+	t, err := template.New("policy-def").Parse(string(defRaw))
+	if err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing the operation policy definition \"%v\" into go template: %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2205,
+		})
+		return Policy{}, err
+	}
+
+	var out bytes.Buffer
+	err = t.Execute(&out, policy.Parameters)
+	if err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing operation policy definition \"%v\": %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2206,
+		})
+		return Policy{}, err
+	}
+
+	def := PolicyDefinition{}
+	if err := yaml.Unmarshal(out.Bytes(), &def); err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing formalized operation policy definition \"%v\" into yaml: %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2207,
+		})
+		return Policy{}, err
+	}
+
+	// update templated policy itself and return, not updating a pointer to keep the original template values as it is.
+	policy.Parameters = def.Definition.Parameters
+	policy.Action = def.Definition.Action
+
+	return policy, nil
+}
+
+// PolicySpecification holds policy specification from ./Policy/<policy>.yaml files
+type PolicySpecification struct {
+	Type    string `yaml:"type" json:"type"`
+	Version string `yaml:"version" json:"version"`
+	Data    struct {
+		Name              string   `yaml:"name"`
+		ApplicableFlows   []string `yaml:"applicableFlows"`
+		SupportedGateways []string `yaml:"supportedGateways"`
+		SupportedAPITypes []string `yaml:"supportedApiTypes"`
+		MultipleAllowed   bool     `yaml:"multipleAllowed"`
+		PolicyAttributes  []struct {
+			Name            string `yaml:"name"`
+			ValidationRegex string `yaml:"validationRegex,omitempty"`
+			Type            string `yaml:"type"`
+			Required        bool   `yaml:"required"`
+		} `yaml:"policyAttributes"`
+	}
+}
+
+// PolicyDefinition holds the content of policy definition which is rendered from ./Policy/<policy>.gotmpl files
+type PolicyDefinition struct {
+	Definition struct {
+		Action     string                 `yaml:"action"`
+		Parameters map[string]interface{} `yaml:"parameters"`
+	} `yaml:"definition"`
+	RawData []byte `yaml:"-"`
+}
```

Comment:
Shall we move these to the top. Maybe after constants. So then when people don't want to know any of the impl details, they can just read these as soon as they open the file.

Reply:
fixed with https://github.com/wso2/product-microgateway/pull/2656/commits/321b5c37e580d65072fbcd8902e950a87217755a